src/eve/datamodels/core.py

```diff
@@ -0,0 +1,1339 @@
+
+    if isinstance(type_annotation, TypeVar):
+        return (
+            _make_type_converter(type_annotation.__bound__, name)
+            if type_annotation.__bound__
+            else toolz.identity
+        )
+
+    if type_annotation is Any:
+        return toolz.identity
+
+    origin_type = xtyping.get_origin(type_annotation)
+
+    if (
+        origin_type is xtyping.Union
+        and type(None) in (args := xtyping.get_args(type_annotation))
+        and len(args) == 2
+    ):
+        # Optional type
+        _inner_type_converter: TypeConverter[_T] = _make_type_converter(args[0], name)
+
+        return cast(TypeConverter[_T], lambda x: x if x is None else _inner_type_converter(x))
+
+    if xtyping.is_actual_type(origin_type):
+        return _make_type_converter(origin_type, name)
+
+    raise exceptions.EveTypeError(
+        f"Automatic type coertion for {type_annotation} types is not supported."
+    )
+
+
+_CACHE_HASH_THRESHOLD: Final = 6
+_KNOWN_MUTABLE_TYPES: Final = (list, dict, set)
+_FROZEN_COLLECTIONS_CHANGES: Final = {
+    list: tuple,
+    set: frozenset,
+    dict: type_definitions.frozendict,
+}
+
+
+def _make_datamodel(  # noqa: C901  # too complex but still readable and documented
+    cls: Type[_T],
+    *,
+    repr: bool,  # noqa: A002   # shadowing 'repr' python builtin
+    eq: bool,
+    order: bool,
+    unsafe_hash: bool,
+    frozen: bool | Literal["strict"],
+    match_args: bool,
+    kw_only: bool,
+    slots: bool,
+    coerce: bool,
+    generic: bool | Literal["True_no_checks"],
+    type_validation_factory: Optional[FieldTypeValidatorFactory],
+    _stacklevel_offset: int = 0,
+) -> Type[_T]:
+    """Actual implementation of the Data Model creation.
+
+    See :func:`datamodel` for the description of the parameters.
+
+    """
+    mro_bases: Tuple[Type, ...] = cls.__mro__[1:]
+
+    if "__annotations__" not in cls.__dict__:
+        cls.__annotations__ = {}
+    annotations = cls.__dict__["__annotations__"]
+    resolved_annotations = xtyping.get_partial_type_hints(cls)
+    annotations_with_extras = xtyping.get_partial_type_hints(cls, include_extras=True)
+
+    frozen, strict_frozen = (True, True) if frozen == "strict" else (frozen, False)
+
+    # Create attrib definitions with automatic type validators and converters
+    # for the annotated fields. The keys in the original annotations are used for
+    # iteration, since the resolved annotations also contain superclasses' annotations
+    for key in annotations:
+        type_hint = annotations[key] = resolved_annotations[key]
+        if xtyping.get_origin(type_hint) is not ClassVar:
+            if xtyping.get_origin(annotations_with_extras[key]) == xtyping.Annotated:
+                _, *type_extras = xtyping.get_args(annotations_with_extras[key])
+            else:
+                type_extras = []
+
+            coerce_field = coerce or _COERCED_TYPE_TAG in type_extras
+            qual_key = f"{cls.__name__}.{key}"
```

Comment:
Is this unique enough? Or is the assumption that the factory is per-class anyways?

Reply:
`qual_key` is only used as argument for the error messages in type validators and converters. I'll rename it to something else to make this clear.